python/ray/data/datasource/partitioning.py

```diff
@@ -0,0 +1,499 @@
+
+    def _parse_dir_path(self, dir_path: str) -> Dict[str, str]:
+        """Directory partition path parser.
+
+        Returns a dictionary mapping directory partition keys to values from a
+        partition path of the form "{value1}/{value2}/..." or an empty dictionary for
+        unpartitioned files.
+
+        Requires a corresponding ordered list of partition key field names to map the
+        correct key to each value.
+        """
+        dirs = [d for d in dir_path.split("/") if d]
+        field_names = self._scheme.field_names
+        assert not dirs or len(dirs) == len(field_names), (
+            f"Expected {len(field_names)} partition value(s) but found "
+            f"{len(dirs)}: {dirs}."
+        )
+        return {field_names[i]: d for i, d in enumerate(dirs)} if dirs else {}
+
+    def __call__(self, path: str) -> Dict[str, str]:
+        """Parses partition keys and values from a single file path.
+
+        Args:
+            path: Input file path to parse.
+        Returns:
+            Dictionary mapping directory partition keys to values from the input file
+            path. Returns an empty dictionary for unpartitioned files.
+        """
+        dir_path = self._dir_path_trim_base(path)
+        if dir_path is None:
+            return {}
+        return self._parser_fn(dir_path)
+
+
+class PathPartitionFilter:
+    """Partition filter for path-based partition formats.
+
+    Used to explicitly keep or reject files based on a custom filter function that
+    takes partition keys and values parsed from the file's path as input.
+    """
+
+    def __init__(
+        self,
+        path_partition_parser: PathPartitionParser,
+        filter_fn: Callable[[Dict[str, str]], bool],
+    ):
+        """Creates a new path-based partition filter based on a parser.
+
+        Args:
+            path_partition_parser: The path-based partition parser.
+            filter_fn: Callback used to filter partitions. Takes a dictionary mapping
+                partition keys to values as input. Unpartitioned files are denoted with
+                an empty input dictionary. Returns `True` to read a file for that
+                partition or `False` to skip it. Partition keys and values are always
+                strings read from the filesystem path. For example, this removes all
+                unpartitioned files:
+                ``lambda d: True if d else False``
+                This raises an assertion error for any unpartitioned file found:
+                ``def do_assert(val, msg):
+                    assert val, msg
+                  lambda d: do_assert(d, "Expected all files to be partitioned!")``
+                And this only reads files from January, 2022 partitions:
+                ``lambda d: d["month"] == "January" and d["year"] == "2022"``
+        """
+        self._parser = path_partition_parser
+        self._filter_fn = filter_fn
+
+    @property
+    def parser(self) -> PathPartitionParser:
```

Comment:
It's currently only used by tests, but I'd expect it to also be useful to anyone constructing a filter since it provides the only path back to the properties of the underlying partition scheme (e.g. via filter.parser.scheme.* to retrieve the base directory, partition field names, etc.).